[Bongf] 미션 2 : 체스판 생성

bong6981 opened the pull request:
안녕하세요! 코드스쿼드 백엔드 Bongf입니다. 
- 지난 피드백을 받고 merge되어 피드백 반영한 내용도 함께 작성했습니다 
- 코드를 작성한 다음 다른 분들이 받은 피드백을 참고하여 코드를 수정했습니다 (커밋 메시지 Fix로 시작되는 부분입니다) 
- Pawn에 다른 객체가 추가되지 않도록 한다 부분은 new Integer("7")로 수행하여 컴파일 에러가 발생하는지 확인하였고 이에 대한 코드는 삭제하였습니다. 
- 가장 어려웠던 부분은 test끼리는 독립적이어야한다는 부분이었습니다. 
    -  학습한 바로는 하나의 테스트당 하나의 로지컬한 컨셉만 테스트하게하라 였는데
    -  board에 pawn을 생성하고 board의 size를 확인하는 부분과 인덱스로 pawn을 찾는 부분이 한 번에 수행되어야 했기 때문에 다른 분들이 받은 피드백을 보고 assertAll로 바꾸었으나 이것이 어쨌든 다른 기능을 확인하는 부분인데 분리해줘야 하지 않을까 고민했습니다 
    - 또, 이전까지는 ParameterizedTest로 돌렸을 때는 white, black 같이 각각의 test 결과가 나왔는데, 이를 없애니 이를 확인할 수 없어 고민했습니다. 
- 워낙 많이 부족하고 아예 개념을 잘못이해하서 삽질하는 고민이 아닌가 하는 생각이 들었지만 나름 긴 시간 동안 고민했는데 답을 찾지 못해 해당 사항을 남겨둡니다.   
코드 리뷰 감사하다는 말씁 드립니다!

bong6981 committed:
Reflect on Feedback

피드백 반영하여 Pawn 클래스 내 필드 private으로 수정
(잘못된 커밋 수정) 피드백 반영하여 매개 변수가 사용될 수 있도록 수정
다른 색깔이 생성 되면 안 될 것 같아 Pawn클래스 안에서 enum으로
색을지정해주고 발생하는 예외에 대해서는 PawnTest에서 예외처리함

bong6981 committed:
Fix incorrect code in enum Color

enum 열거자는 대문자로 작성된다는 조건을 놓쳐서 다시 대문자로 수정
그에 따라 생성자와, getColor에서 대소문자 수정하는 코드 작성

bong6981 committed:
Modify Pawn.class

-색이 없는 Pawn을생성하는 테스트createPawn()을 추가했고, 그에 따라 Pawn
클래스에도 기본 생성자 추가
-test에도 상수를 활용하라는 조건이 있어서 Pawn의 내부클래스로 있던 enum을
별도의 클래스로 빼줌
-상수 활용을 위해 string을 매개별수로 넣어주었던
createPawnWithColor메서드도 매개변수를 enum 클래스 Color의 객체를 받게함
-그에 따라 Pawn의 생성자 매개변수도 Color로 수정하고 getColor()의
반환값도 Color 객체로 수정함

bong6981 committed:
Add BoardTest

-BoardTest를 추가하고 Board 클래스 추가
-BoardTest에서 Pawn을 생성하고, (생성한 Board board에 대해) board.add()를 BoardTest에 작성 후 Board에도 add 메서드
추가
	-Board에서 추가한 Pawn을 담아 줄수 있는 List 생성, add메서드 실행시
해당 Pawn을 List(list명 board)에 추가하도록 구현
-BoardTest에서 생성한 board에 대해 보드에 올라간 pawn의 개수를 반환하는 board.size() 추
	-Board에서 생성한 List board의 크기를 반환하도록 size()메서드
	구현
-BoardTest애서 생성된 board에 대해 추가된 Pawn을 담고 있는 List의
인덱스로 해당 Pawn을 찾는 findPawn(int)츷작성
	-Board에서 파라미터로 전달된 인덱스에 해당하는 Pawn을 List board에서 찾아 반환하는 findPawn()메서드 구현

bong6981 committed:
Seperate package

-chess 패키지 새로 만들어 Board와 BoardTest를 해당 패키지로 옮김
-컴파일에러가 발생하지 않도록 다른 패키지에 있는 것들을 import해줌
-default였던 메소드나 클래스를 다른 패키지에서 사용할 수 있도록
public으로 변경

bong6981 committed:
Remove redundant codes

- 흰, 검은 Pawn을 생성하고 확인하는 코드가 겹처 중복제거
- Pawn을 생성하고 확인하는 부분을 createPawnAndVerify 메서드로 따로 뺌
- board를 클래스 멤버로 선언
- 추가한 Pawn의 수를 계산해주는 count 멤버변수를 추가해서
  assertThatt실행할 때 pawn의 개수로 값을 확인할 수 있게 함

bong6981 committed:
Fix the Code 1 : Delete ParameterizedTest

- 다른 멤버분들의 코드에 대한 코드 리뷰를 보고 내 코드도 수정했다.
- ParamterizedTest를 사용ㅎ지 말라는 feedback을 보고 삭제했다.

bong6981 committed:
Fix the Code 2 : Change access modifiers in test code

- 다른 멤버분들의 코드에 대한 코드 리뷰를 보고 코드를 수정했다.
- test 코드에 접근제어자가 public이 아니어도 된다는 말에 default로
  수정했다 ( 최소한으로 하기 위해 )

bong6981 committed:
Fix the Code 3 : Seperate function

- 한 메소드에pawn을 생성하고 배열에 add되었는지 확인하는 두 가지 기능이
  있어 기능을 나눔
  - createPawnOntheBoard에서 생성후 해당폰을
    addPawnOnTheBoardAndVerify로 넘김
- test는 독립적으로 진행되어야 한다는 강의를 듣고, 공유되는 멤버 변수를
  줄이면 좋을 것 같아 count를 삭제하고 정수로 직접적인 값을 넘겨줌

bong6981 committed:
Fix the Code 4 : Change to assertAll

- 다른 멤버분들의 코드에 대한 코드 리뷰를 보고 코드를 수정했다.
- 'assertions을 여러 개나열하면 하나의 검증이 fail 되면 더이상 다른 assertions을 실행하지 않는다. 어디서 fail했는지 알기 위해선 assertAll을 써야한다. 어느 하나가fail하더라도 모든 assertions을 검증하기 때문에 어디서 fail이 발생했는지 (다른 assertions은 문제가 없는 것인지) 한 번에 알 수 있다'라는 내용을 학습하고 assertAll로 고쳤다.

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public class Board {
+    private List<Pawn> board = new ArrayList<>();
```
pawn list 멤버 변수 이름으로 board 는 맞지 않는 것 같습니다. 이미 Board 클래스의 인스턴스에도 board라는 이름을 쓸 것인데 (테스트코드를 보면..) 

```suggestion
    private List<Pawn> pawns = new ArrayList<>();
```
로 수정해주세요. ㅎㅎ

bong6981 replied:
네! 제가 해당 미션의 내용을 잘못이해하여 멤버 변수 명을 잘못 지은 것 같습니다.  말씀 주신 사항으로 수정했습니다!

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,22 @@
+public class Board {
+    private List<Pawn> board = new ArrayList<>();
+
+    void add(Pawn pawn) {
+        board.add(pawn);
+    }
+
+    int size() {
+        return board.size();
+    }
+
+    Pawn findPawn(int index) {
+        return board.get(index);
+    }
```
접근제어자를 default로 두신 이유가 있으신지요?

bong6981 replied:
우선 접근제어자는 최대한 좁게 설정하고 필요시 확장하라고 학습하여 학습내용을 적용해보았습니다! 아직 Board의 메서드들이 다른 패키지에서는 쓰이지 않아 default로 두었습니다!

LarryJung commented on src/main/java/pieces/Pawn.java:
```diff
@@ -1,13 +1,17 @@
 
 public class Pawn {
-    String color;
+    private final Color color;
```
final 키워드를 붙이면 어떤 장점이 있을까요? (단점도 있을까요?)

bong6981 replied:
final 키워드를 붙이면 상속 및 재정의를 막을 수 있다 (변경을 막을 수 있다)라고 학습했습니다. 여기서는 상수 정의에 사용했기 때문에 생성시에 색을 할당했다면 재할당이 불가능한 것으로 이해하고 적용했습니다. 장점은 다른 곳에서 색을 바꾸려고 할 때 막을 수 있다는 점이고 단점은 혹 코드를 작성하다가 색을 바꿔야 하는 기능들을 구현할 수 없다는 단점을.. 생각해 보았습니다!

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+        board = new Board();
+    }
+
+    void addPawnOnTheBoardAndVerify(Pawn pawn, int i) {
+        board.add(pawn);
+        assertAll(
+                () -> assertThat(board.size()).isEqualTo(i),
+                () -> assertThat(board.findPawn(i - 1)).isEqualTo(pawn)
+        );
+    }
```
요 테스트코드는 다른 이유가 아니라면 private으로 선언하는게 나을 것 같네요!

bong6981 replied:
네!  해당 Boardtest 클래스 안에서만 사용되어 private으로 바꿨습니다! 놓치고 있었네요.

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+        board = new Board();
+    }
+
+    void addPawnOnTheBoardAndVerify(Pawn pawn, int i) {
```
메서드 네이밍은 아마 평생.. 고민이 되실건데 그냥 저라면(저도 아직도 이름정할 때 고민이 많네요..) 
`addPawnVerify` 정도로 하고 넘어가지 않을까 싶기도 하네요 ㅎㅎ

메서드, 변수, 클래스, 인터페이스 등 각종 네이밍 컨벤션은 어떻게 되고 어떻게 지으면 적절할 지 많이 고민해보시고 찾아보시면 도움 많이 되실 거에요

LarryJung replied:
그리고.. 로지컬한 묶음이라는 것은 생각하기 나름일 수도 있습니다. 

pawn 을 추가하는 기능을 테스트한다.
-> 넣었을 때 사이즈가 늘어난다 & 넣은 것을 정확히 찾아낼 수 있다.

요렇게 생각을 하셨다면 특별히 분리할 필요는 없어 보이네요 ㅎㅎ

bong6981 replied:
말씀주신 네이밍 이름이 더 적합한 것 같아 해당 이름으로 변경했습니다. 네이밍 컨벤션에 대해 공부해보겠습니다! 로지컬한 묶음에 대한 고민이 많았는데 이렇게 생각하면 되는 군요! 감사합니다

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+        board.add(pawn);
+        assertAll(
+                () -> assertThat(board.size()).isEqualTo(i),
+                () -> assertThat(board.findPawn(i - 1)).isEqualTo(pawn)
```
제 생각엔 테스트코드에 '로직'이 들어간 느낌인데요! 
`findPawn(i - 1)` 이 1을 넣으면 0번째것을 꺼내는 것을 보니..  빼기 1 해주는 '로직'은 findPawn 메서드 안으로 들어가야 더 적절할 것으로 보입니다.

bong6981 replied:
이 피드백을 받고 test 코드에 로직을 넣으면 그 또한 버그를 만들어 낼 수 있기 때문에 지양해야 한다는 사실을 알았습니다.  

다만 미션에서 findPawn 메서드 자체가 미션의 'Pawn을 추가할 때마다 목록을 얻어와 추가한 Pawn을 정상적으로 가지는 것을 확인한다.' 부분에서  해당 인덱스를 넣고 리스트에 pawn이 잘 들어가 있는지 확인해야 한다고 이해했습니다. 해서 새로 커밋한 바와 같이 아예 number(폰의 갯수), index(list에서 pawn이 들어가야 할 자리)로 두 개의 매개변수를 받아 넘겨주는 것으로 수정했는데 어떠실지 문의드립니다.

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Test
+    @DisplayName("흰색, 검은색 Pawn을 생성해 board에 추가한다")
+    void createPawnOnTheBoard() throws Exception {
```
throws Exception은 어떻게 붙게 되었을까요?

bong6981 replied:
생각없이 예제 코드를 따라하다 보니 해당 부분이 들어간 것 같습니다. ㅠㅠ
-Exception은 모든 예외의 최고조상이다.
-보통 메서드에 예외를 선언할 때에는 메서드 내에서 발생할 수 있는 (일반적으로 필수처리예외만 적는다) 예외를 명시해준다. 
-메서드를 사용하려는 사람들은 이 메서드의 선언된 예외를 보고 어떤 예외가 발생할지 예측할 수 있다. 
라고 다시 학습했습니다. 
다만, 지금 이 코드에서 제가 생각할 수 있는 필수처리 예외를 찾기 어려워 삭제했습니다!

bong6981 replied:
학습 속도가 느리고 아직 기초도 많이 부족한 편이라 주신 피드백을 보고 학습하고 적용하는 것까지 시간이 오래 걸렸습니다. 자세한 피드백 주셔서 감사합니다!

LarryJung replied:
넵 ㅎㅎ 자바에서는 CheckedException, UnCheckedException 두 부류가 있는데 차이점에 대해서 학습해 보시는 것을 권해드려요~

LarryJung reviewed (commented):
수고하셨습니다. 요구사항대로 작 작성해주셨지만, 보면서 질문들을 몇가지 남겨드렸네요.^^ 확인 부탁드리고 답글로 소통하시지요!

honux77 reviewed (changes requested):
리뷰어 코멘트 보고 개선해 주세요~

bong6981 committed:
수정사항 반영

수정사항 반영하여 코드를 수정
- 1. Board 클래스의 리스트명을 board에서  pawns로 변경
- 2. test클래스 내부에서만 사용하는 메서드의 접근제어자를 private으로
  변경
- 3. 추천받은 메서드 명으로 addPawnVerify로 변경
- 4. pawn의 갯수를 매개변수로 받아 -1을 하여 list에서 해당 폰을 찾는
  과정의 수정
  -	추천받은 방법은 findPawn 자체에 로직을 변경하여 같은 매개변수를
	넣고 test 코드에서 로직을 삭제하는 것
  -	pawn을 찾는다는 의미 자체가 해당 인덱스로 찾아야 할 의미는
	살려야 할 것 같아서 매개변수를 추가하여 pawn을 찾는 정수를 따로
	받아줌

bong6981 committed:
수정사항 반영

- 수정 사항 반영 부분 중 메서드명 수정이 반영 되지 않아 다시 커밋

bong6981 commented:
안녕하세요. 학습 속도가 느리고 아직 기초도 많이 부족한 편이라 주신 피드백을 보고 학습하고 적용하는 것까지 시간이 오래 걸렸습니다. 자세한 피드백 주셔서 감사합니다!

말씀 주신 피드백 반영했습니다. 
 - 미션 내용 자체를 잘못 이해해  잘못  네이밍한 변수명 board를 pawns로 수정
 - BoardTest 내에서만 쓰이는 메서드에 대해 private으로 선언 
 - 추천주신 메서드명(addPawnVerify)으로 수정, 네이밍 컨벤션 더 학습 필요 
 - 이해 없이 작성한 코드 throws Exception을 삭제. 
     - 현재에는 처리할 예외를 발견하지 못해 해당 부분 삭제 
 - test에 로직이 들어가는 것을 지양해야 한다는 것을 배웠으며, 이에 따라 test안의 로직 삭제 
     - 다만 findPawn() 자체가 해당 리스트에 pawn이 잘 들어갔는지 확인하는 기능이라 해당 인덱스로 찾아야 한다고 생각하여 인덱스에 해당하는 부분을 매개변수로 넣어줌 
 
그 외 질문주신 내용 
- Board클래스의 접근제어자를 default로 설정한 이유
    -  학습과정에서 최소한으로 접근 제어자를 설정해두고 필요시 확장한다고 배워, 다른 패키지에서 아직 사용된 일이 없어 디폴트 패키지로 설정 
- final 키워드를 붙인 장점 
   -  한 번 값이 할당되면 다른 곳에서 변경할 수 없도록 방지 
   -  단점은 혹시 색상을 바꿔야할 기능이 필요하다면 이 기능을 위해 다른 멤버 변수를 선언하는 등 다른 방법으로 접근해야 함 (검색으로 찾지 못해 위와 같이 최대한 생각해보려고 노력해보았습니다) 
으로 정리했습니다.

honux77 commented:
최소한의 접근 제어자 = private 입니다 ㅎㅎ. 특별한 이유 없으면 클래스는 public, 멤버는 private을 염두에 두세요.
